src/rgw/rgw_auth_s3.cc

```diff
@@ -197,16 +197,15 @@ bool rgw_create_s3_canonical_header(const req_info& info,
   if (qsr) {
     date = info.args.get("Expires");
   } else {
-    const char *str = info.env->get("HTTP_DATE");
+    const char *str = info.env->get("HTTP_X_AMZ_DATE");
     const char *req_date = str;
-    if (str) {
-      date = str;
-    } else {
-      req_date = info.env->get("HTTP_X_AMZ_DATE");
+    if (str == NULL) {
+      req_date = info.env->get("HTTP_DATE");
       if (!req_date) {
         dout(0) << "NOTICE: missing date for auth header" << dendl;
         return false;
       }
+      date = req_date;
```

Comment:
@joke-lee that example request corresponds to this case where `str != NULL` though, which causes us to use an empty `date` in `StringToSign`. and this is the correct behavior:
> if you include the `x-amz-date` header, use the empty string for the `Date` when constructing the `StringToSign`

i think the example itself is incorrect. when given that request, the code in this pr produces this `StringToSign`, instead:
> DELETE\n
\n
\n
\n
x-amz-date:Tue, 27 Mar 2007 21:20:26 +0000\n
/johnsmith/photos/puppy.jpg

i wrote an extra s3test https://github.com/ceph/s3-tests/pull/171 to verify that. `test_object_create_date_and_amz_date` fails against master, and passes with this pr 👍